CSS Class Names: Add dropdown to choose from block styles

ramonjd opened the pull request:
## Description

Resolves https://github.com/WordPress/gutenberg/issues/33934

These changes allow us to switch between block styles beside the advanced controls, and offer a quick way to select or reinstate block styles classes.

<img width="400" alt="the gif" src="https://user-images.githubusercontent.com/6458278/141220055-a74181a8-d254-4ead-b8c4-9016d675d23f.gif">


<img width="400" alt="Screen Shot 2021-11-11 at 12 38 42 pm" src="https://user-images.githubusercontent.com/6458278/141221594-30d0e9f5-8469-4460-9e57-cd7fd55dcd52.png">

## Testing

Please make yourself comfortable. 

Insert a block that has block styles, for example the Image, Table or Buttons Block in TT1 Blocks theme.

Check that:
- [ ] Selecting a block style from the **Advanced > Additional CSS class(es) field** dropdown changes the block style.
- [ ] Switching block styles using the block styles preview buttons also updates the **Advanced > Additional CSS class(es) field**
- [ ] Switching block styles using the block styles preview buttons also updates the **Advanced > Additional CSS class(es) field**
- [ ] Copying a valid block style classname into the field, and deleting it, updates the block styles accordingly
- [ ] Deleting a block style classname from the field causes the block style to return to the "Default" style
- [ ] The dropdown doesn't appear for blocks that have no block styles, e.g., a Paragraph Block in TT1 Blocks theme.
- [ ] Things look and work okay in narrow viewports.

Check TextControl in Storybook for a new story (`withChildren`);

<img width="454" alt="Screen Shot 2021-12-08 at 11 27 22 am" src="https://user-images.githubusercontent.com/6458278/145126900-c6803588-f7ee-4207-8109-006a78aa7067.png">



## Types of changes
Additional functionality.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ramonjd committed:
initial commit

ramonjd committed:
Adding dispatch to update block styles

ramonjd commented:
We might need the `getRenderedStyles` method introduced in https://github.com/WordPress/gutenberg/pull/34522 so that we show "Default" where one exists:

<img width="276" alt="Screen Shot 2021-09-20 at 3 27 04 pm" src="https://user-images.githubusercontent.com/6458278/133960101-40ba5ff9-b1b7-46d0-9aa4-5732b0557866.png">

Also, alignment of the icon is a little tricky here. 

We'd want to align the ellipsis icon with the top (the Additional CSS class(es) label) or with the input field. 

But because we're using `<TextControl />`, which encompasses a label and input control, the most straightforward approach is to align the icon to the top. 

It does look much neater in the designs at #33934 – where the ellipsis icon is aligned with the input field – but in order to achieve this we might have to either:

1. extend `<TextControl />` to accept children (so we can render them as siblings of the input, or 
2. roll our own. 

No 2 runs the risk of losing standardization since all items in the Advanced panel use this component. 

No 1 would have be thoroughly tested given the frequency with which `<TextControl />` is used in both web and native platforms.

ramonjd committed:
Formatting and adding a custom classname to control the icon container margin

ramonjd committed:
Some light formatting/refactoring
Eyeballing extra margins
Ensuring the additional class input field stretches in mobile

shaunandrews committed:
Update the menu heading to "Existing Styles."

shaunandrews committed:
Switching from flex to absolute positioning for the dropdown button. This might be a terrible idea.

shaunandrews commented:
I pushed a few updates. First I updated the menu label from "Block style classes" to "Existing styles." Then, I decided to be reckless and switch from `flex` to `position: absolute;` — this may be a mistake, but I was able to get it looking like this:

<img width="271" alt="image" src="https://user-images.githubusercontent.com/191598/141837526-b4ad9a8f-7e80-413f-a1d9-590a734c5671.png">

Otherwise, this PR seems to work as expected; The input respects the styles from the accordion above; I'm able to choose from the menu to switch styles; Typing in a class manually correctly selects the existing class.

ramonjd committed:
This commit tests embedding the dropdown as a child of the TextControl component so we can position the icon more accurately.
We're also adding a background color to the icon in case the underlying text flow beneath it.
Finally, in the case of smaller screens, we're bumping up the size of the icon.

ramonjd commented:
Thanks for the idea @shaunandrews I think we can work with it.

I've taken it a bit further in fa222296b9ab65e4834ab0e93b46bdbf9c6b0388 by passing down the dropdown as a child to `<TextControl />` , and also taking care of the following scenarios:

Input overflow:
<img width="272" alt="Screen Shot 2021-11-16 at 9 30 00 am" src="https://user-images.githubusercontent.com/6458278/141866654-52fd89fb-7c65-4ac3-beaa-cee0b32ce9c6.png">

Small screens:
<img width="272" alt="Screen Shot 2021-11-16 at 9 23 11 am" src="https://user-images.githubusercontent.com/6458278/141866657-b311833a-35d5-4460-b1a9-a90f08734f47.png">

Label overflow:
<img width="272" alt="Screen Shot 2021-11-16 at 9 30 26 am" src="https://user-images.githubusercontent.com/6458278/141866647-b3e4c696-f852-4f9a-a63a-6287ffde8e5b.png">


Passing down the dropdown as a child to `<TextControl />` means we can use the text control's parent as the relative positioning reference, and position from the bottom where there's less chance of variable height.

If you think this approach is okay, I might need to split this PR and do that bit first since I'd have to update the stories and promote the benefits 😄 

Result:
<img width="466" alt="Screen Shot 2021-11-16 at 10 00 18 am" src="https://user-images.githubusercontent.com/6458278/141866798-7aa125c3-8d58-47a8-b6e0-2f4c52f8aa6e.png">

<img width="272" alt="Screen Shot 2021-11-16 at 10 00 32 am" src="https://user-images.githubusercontent.com/6458278/141866790-15bb55bc-fe65-497d-87c3-2ef2a8a2710d.png">

glendaviesnz commented on packages/block-editor/src/hooks/custom-class-name.js:
```diff
@@ -55,27 +94,111 @@ export const withInspectorControl = createHigherOrderComponent(
+									) }
+								>
+									{ hasBlockStyles && (
+										<DropdownMenu
```
Would it be better to split this out into its own component? That might help reduce the depth of component nesting we are ending up with here.

ramonjd replied:
Thanks for the 👀 

You're making sense. I like it.

ramonjd replied:
fd15e64f5cf4f2a066c9da97cb34073879cd98f6 splits out the custom inspector controls into a component 👍

ramonjd committed:
Extracting inspector controls component into a component.

ramonjd committed:
Extracting dropdownmenu component

shaunandrews commented:
<img width="269" alt="image" src="https://user-images.githubusercontent.com/191598/144505008-18d0d25f-f7bc-4419-9b06-0d7eacc58043.png">

There's a slight mis-alignment (1px too high, I think), but otherwise this seems good.

> I might need to split this PR and do that bit first since I'd have to update the stories and promote the benefits

Seems like a solid step forward.

ramonjd commented:
Thanks for testing @shaunandrews 🙇 

> There's a slight mis-alignment (1px too high, I think), but otherwise this seems good.

Pah! 1px, 2px...  :trollface: 

Out of interest, which browser are you using? Here's Chrome, Safari and Firefox for me, in order: 

<img width="258" alt="Screen Shot 2021-12-03 at 11 41 26 am" src="https://user-images.githubusercontent.com/6458278/144525468-1975f9c8-d2ca-46e4-a16b-5b5d7a53c223.png">

<img width="276" alt="Screen Shot 2021-12-03 at 11 42 27 am" src="https://user-images.githubusercontent.com/6458278/144525465-a2732691-f56d-4df3-94ad-77b5aefb8570.png">
<img width="266" alt="Screen Shot 2021-12-03 at 11 43 14 am" src="https://user-images.githubusercontent.com/6458278/144525462-f1a24204-5ed6-4796-9064-af60c46fc943.png">

shaunandrews commented:
Safari. I may or may not have zoomed in so my old eyes can see things.

ramonjd commented:
> I may or may not have zoomed in so my old eyes can see things.

Good call. When I zoom in or out by various factors, it's visible.

<img width="331" alt="Screen Shot 2021-12-03 at 11 48 36 am" src="https://user-images.githubusercontent.com/6458278/144525855-02808917-e857-48a4-b8ec-779ad9c4918a.png">

ramonjd commented on packages/components/src/text-control/index.js:
```diff
@@ -65,6 +66,7 @@ function TextControl(
 				ref={ ref }
 				{ ...props }
 			/>
+			{ children }
```
By way of explanation, rendering `children` here was a trade off. 

Following the design requirements, we need to be able to add a dropdown menu as a sibling of the text control input.

We could have created a custom component using the constituent parts of `<TextControl />`. The Advanced panel, however, uses `<TextControl />` multiple times, which means that updates would have to be copied over to the custom component to retain consistency. 

Does adding `children` here contravene any component guidelines from a technical or philosophical standpoint? 

As far as I can tell, there are no side effects to adding and using the `children` prop, though I am worried that it might be "impure" in that it invites manifestations of a base component. 

I'm overthinking things, yes.

ramonjd replied:
I'd be grateful for your thoughts here @ciampo whenever you get time. 🙏

ciampo replied:
Thank you for the ping! Always happy to advise / give direction on components :)

I would say that rendering `children` inside a `TextControl` goes a bit against the way the component is supposed to work. `TextControl`, in my mind, is supposed to be an enriched `input` field.

I would personally recommend that, instead of adding `children` to `TextInput`, you render the dropdown menu as a sibling of `TextInput` in the `CustomClassNameControl` component.

Also, cc'ing @diegohaz and @mirka in case they wanted to add their views here as well

ramonjd replied:
Thanks for the feedback and for confirming what the tiny voice in my brain was trying to tell me. I was truly struggling with the conflict 😆 

> I would personally recommend that, instead of adding children to TextInput, you render the dropdown menu as a sibling of TextInput in the CustomClassNameControl component.

I'll try that out, cheers! If the other Advanced control inputs end up looking or working differently I can try to switch them all over to TextInput as well.

ramonjd replied:
Ah, I just realized that `TextInput` isn't a component yet. Or at least we haven't migrated it from g2.

I should know, I started it and [never finished](https://github.com/WordPress/gutenberg/pull/32381)! 😅

ciampo replied:
> Ah, I just realized that TextInput isn't a component yet. Or at least we haven't migrated it from g2.

We did initially look into it as well after you started it, but we later stopped because it was a very complicated task, since `TextInput` is very complex (it tries to handle single/multi line text and number input) and represents a very different approach to what we currently have in Gutenberg.

ramonjd replied:
> We did initially look into it as well after you started it, but we later stopped because it was a very complicated task

Thanks for the background info. 🙇 

I discovered personally how tricky it was 😬  so I feel better that folks with a lot more experience than I have also felt the same!

ramonjd committed:
TextControl:
- Added prop notes for children in readme
- Added story for withChildren

ramonjd commented:
@shaunandrews How would you feel if I parked this one for now? I'm not sure how to achieve the current design without re-creating TextControl, which seems like an overkill.

ramonjd commented:
Closing for now until we can revisit the approach. 

We can always reopen. 

Thanks to everyone for their help on this one!
